graph/path/control_flow.go

```diff
@@ -4,169 +4,187 @@
-			lt.bucket[lt.parent[wid].ID()].Remove(v)
+		for v := range w.parent.bucket {
+			delete(w.parent.bucket, v)
+
 			u := lt.eval(v)
-			if lt.semi[u.ID()] < lt.semi[vid] {
-				lt.dom[vid] = u
+			if u.semi < v.semi {
+				v.dom = u
 			} else {
-				lt.dom[vid] = lt.parent[wid]
+				v.dom = w.parent
 			}
 		}
-
 	}
 
 	// step 4.
-	for i := 2; i <= lt.n; i++ {
-		w := lt.vertex[i]
-		wid := w.ID()
-		if lt.dom[wid].ID() != lt.vertex[lt.semi[wid]].ID() {
-			lt.dom[wid] = lt.dom[lt.dom[wid].ID()]
+	for _, w := range lt.nodes[1:] {
+		if w.dom.node.ID() != lt.nodes[w.semi].node.ID() {
+			w.dom = w.dom.dom
 		}
 	}
-	delete(lt.dom, root.ID())
 
-	nodes := make(set.Nodes)
-	for _, n := range g.Nodes() {
-		nodes.Add(n)
-	}
+	// Construct the public-facing dominator tree structure.
+	dominatorOf := make(map[int64]graph.Node)
 	dominatedBy := make(map[int64][]graph.Node)
-	for nid, d := range lt.dom {
-		did := d.ID()
-		dominatedBy[did] = append(dominatedBy[did], nodes[nid])
+	for _, w := range lt.nodes[1:] {
+		dominatorOf[w.node.ID()] = w.dom.node
+		did := w.dom.node.ID()
+		dominatedBy[did] = append(dominatedBy[did], w.node)
 	}
-	return DominatorTree{root: root, dominatorOf: lt.dom, dominatedBy: dominatedBy}
+	return DominatorTree{root: root, dominatorOf: dominatorOf, dominatedBy: dominatedBy}
 }
 
+// lengauerTarjan holds global state of the Lengauer-Tarjan algorithm.
+// This is a mapping between nodes and the postordering of the nodes.
 type lengauerTarjan struct {
-	// The vertex which is the parent of vertex w
-	// in the spanning tree generated by the search.
-	parent map[int64]graph.Node
+	// nodes is the nodes traversed during the
+	// Lengauer-Tarjan depth-first-search.
+	nodes []*ltNode
+	// indexOf contains a mapping between
+	// the id-dense representation of the
+	// graph and the potentially id-sparse
+	// nodes held in nodes.
+	//
+	// This corresponds to the vertex
+	// number of the node in the Lengauer-
+	// Tarjan algorithm.
+	indexOf map[int64]int
+}
 
-	// The set of vertices v such that (v, w) is an edge
-	// of the graph.
-	pred map[int64][]graph.Node
+// ltNode is a graph node with accounting for the Lengauer-Tarjan
+// algorithm.
+//
+// For the purposes of documentation the ltNode is given the name w.
+type ltNode struct {
+	node graph.Node
 
-	// semi[w] is a number defined as follows:
-	// (i)   Before vertex w is numbered, semi[w] = 0, false.
-	// (ii)  After w is numbered but before its semidominator
-	//       is computed, semi[w] is the number of w.
-	// (iii) After the semidominator of w is computed, semi[w]
-	//       is the number of the semidominator of w.
-	semi map[int64]int
+	// parent is vertex which is the parent of w
+	// in the spanning tree generated by the search.
+	parent *ltNode
+
+	// pred is the set of vertices v such that (v, w)
+	// is an edge of the graph.
+	pred []*ltNode
 
-	// vertex[i] is the vertex whose number is i.
-	vertex []graph.Node
+	// semi is a number defined as follows:
```

Comment:
Note that `semi` could be replace with a pair of fields: `number int` and `semi *ltNode`, but the current imlpementation seems simpler.